From: Jann Horn <jannh@google.com>
Subject: slub: Introduce CONFIG_SLUB_RCU_DEBUG
Date: Fri, 09 Aug 2024 17:36:56 +0200

  - move rcu_barrier() out of locked region (vbabka)
  - rearrange code in slab_free_after_rcu_debug (vbabka)

Link: https://lkml.kernel.org/r/20240809-kasan-tsbrcu-v8-2-aef4593f9532@google.com
Tested-by: syzbot+263726e59eab6b442723@syzkaller.appspotmail.com
Reviewed-by: Andrey Konovalov <andreyknvl@gmail.com>
Acked-by: Marco Elver <elver@google.com>
Acked-by: Vlastimil Babka <vbabka@suse.cz> (slab)
Signed-off-by: Jann Horn <jannh@google.com>
Cc: Alexander Potapenko <glider@google.com>
Cc: Andrey Ryabinin <ryabinin.a.a@gmail.com>
Cc: Christoph Lameter <cl@linux.com>
Cc: David Rientjes <rientjes@google.com>
Cc: David Sterba <dsterba@suse.cz>
Cc: Dmitry Vyukov <dvyukov@google.com>
Cc: Hyeonggon Yoo <42.hyeyoo@gmail.com>
Cc: Joonsoo Kim <iamjoonsoo.kim@lge.com>
Cc: Pekka Enberg <penberg@kernel.org>
Cc: Roman Gushchin <roman.gushchin@linux.dev>
Cc: Vincenzo Frascino <vincenzo.frascino@arm.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/slab_common.c |   22 +++++++++++++---------
 mm/slub.c        |    5 ++---
 2 files changed, 15 insertions(+), 12 deletions(-)

--- a/mm/slab_common.c~slub-introduce-config_slub_rcu_debug-v8
+++ a/mm/slab_common.c
@@ -576,15 +576,6 @@ void kmem_cache_destroy(struct kmem_cach
 	if (unlikely(!s) || !kasan_check_byte(s))
 		return;
 
-	cpus_read_lock();
-	mutex_lock(&slab_mutex);
-
-	rcu_set = s->flags & SLAB_TYPESAFE_BY_RCU;
-
-	s->refcount--;
-	if (s->refcount)
-		goto out_unlock;
-
 	if (IS_ENABLED(CONFIG_SLUB_RCU_DEBUG) &&
 	    (s->flags & SLAB_TYPESAFE_BY_RCU)) {
 		/*
@@ -593,10 +584,23 @@ void kmem_cache_destroy(struct kmem_cach
 		 * defer their freeing with call_rcu().
 		 * Wait for such call_rcu() invocations here before actually
 		 * destroying the cache.
+		 *
+		 * It doesn't matter that we haven't looked at the slab refcount
+		 * yet - slabs with SLAB_TYPESAFE_BY_RCU can't be merged, so
+		 * the refcount should be 1 here.
 		 */
 		rcu_barrier();
 	}
 
+	cpus_read_lock();
+	mutex_lock(&slab_mutex);
+
+	rcu_set = s->flags & SLAB_TYPESAFE_BY_RCU;
+
+	s->refcount--;
+	if (s->refcount)
+		goto out_unlock;
+
 	err = shutdown_cache(s);
 	WARN(err, "%s %s: Slab cache still has objects when called from %pS",
 	     __func__, s->name, (void *)_RET_IP_);
--- a/mm/slub.c~slub-introduce-config_slub_rcu_debug-v8
+++ a/mm/slub.c
@@ -4563,9 +4563,8 @@ static void slab_free_after_rcu_debug(st
 		return;
 
 	/* resume freeing */
-	if (!slab_free_hook(s, object, slab_want_init_on_free(s), true))
-		return;
-	do_slab_free(s, slab, object, object, 1, _THIS_IP_);
+	if (slab_free_hook(s, object, slab_want_init_on_free(s), true))
+		do_slab_free(s, slab, object, object, 1, _THIS_IP_);
 }
 #endif /* CONFIG_SLUB_RCU_DEBUG */
 
_
